src/hackyballs/model.js

```diff
@@ -0,0 +1,170 @@
+/* exported HBModelBase, HBModelGlobal, HBModelObject */
+
+const {Gio, GLib, GObject} = imports.gi;
+
+const ViewName = 'view.JSContext.globalParameters';
+const BusName = 'com.endlessm.hackyballs';
+const AppObjectPath = '/com/endlessm/hackyballs';
+const KnowledgeControlIface = `
+<node xmlns:doc="http://www.freedesktop.org/dbus/1.0/doc.dtd">
+  <interface name='com.endlessm.Clippy'>
+    <method name='Set'>
+      <arg type='s' name='object' />
+      <arg type='s' name='property' />
+      <arg type='v' name='value' />
+    </method>
+    <method name='Get'>
+      <arg type='s' name='object' />
+      <arg type='s' name='property' />
+      <arg type='v' name='value' direction='out'/>
+    </method>
+    <method name='Connect'>
+      <arg type='s' name='object' />
+      <arg type='s' name='signal' />
+      <arg type='s' name='detail' />
+    </method>
+    <signal name='ObjectNotify'>
+      <arg type='s' name='object' />
+      <arg type='s' name='property' />
+      <arg type='v' name='value' />
+    </signal>
+  </interface>
+</node>
+`;
+
+const _propFlags = GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT;
+
+var HBModelBase = GObject.registerClass({
+}, class HBModelBase extends GObject.Object {
+    _init(props = {}) {
+        super._init(props);
+        const Proxy = Gio.DBusProxy.makeProxyWrapper(KnowledgeControlIface);
+        this._proxy = new Proxy(Gio.DBus.session, BusName, AppObjectPath);
+    }
+
+    bindProperties(map = {}) {
+        const invertMap = {};
+
+        const props = GObject.Object.list_properties.call(this.constructor.$gtype);
+        props.forEach(prop => {
+            const localProp = prop.get_name();
+            const proxyProp = map[localProp] ? map[localProp] : localProp;
+            this._bindLocalToProxy(localProp, proxyProp);
+            invertMap[proxyProp] = localProp;
+        });
+
+        this._proxy.connectSignal('ObjectNotify', (p, s, [o, proxyProp, ret]) => {
+            const value = ret.deep_unpack();
+            const localProp = invertMap[proxyProp] ? invertMap[proxyProp] : proxyProp;
+
+            /* XXX use GObject.param_values_cmp */
```

Comment:
I think I figured out that `param_values_cmp` and `param_value_validate` are currently unusable from GJS, because you don't have control over what type of GValue gets created, but to use the function correctly, the passed-in GValues must adhere to certain types. And there's no way to express that in g-i annotations.

I think it might be possible to make this function work correctly by overriding `GObject.ParamSpec.prototype.vfunc_values_cmp()` and just reimplementing the logic of the C function in JS, but I agree to not block on that. Would you mind opening a bug for GJS?

Reply:
Yes, ill report it.

Reply:
https://gitlab.gnome.org/GNOME/gjs/issues/210